packages/devtools/lib/src/debugger/console_area.dart

```diff
@@ -54,19 +59,83 @@ class ConsoleArea implements CoreElementView {
     });
   }
 
+  Position _getDocumentEnd() {
+    final doc = _editor.getDoc();
+    final lastLine = doc.lastLine();
+    return Position(lastLine, doc.getLine(lastLine).length);
+  }
+
   void _append(String text) {
-    // append text
-    _editor
-        .getDoc()
-        .replaceRange(text, Position(_editor.getDoc().lastLine() + 1, 0));
-
-    // scroll to end
-    final int lastLineIndex = _editor.getDoc().lastLine();
-    final String lastLine = _editor.getDoc().getLine(lastLineIndex);
-    _editor.scrollIntoView(lastLineIndex, lastLine.length);
+    final scrollInfo = _editor.getScrollInfo();
+    // If the last line of the console is visible, after appending content we
+    // should scroll so that the last line of the editor is still visible.
+    // TODO(jacobr): add an optional setting to always scroll to the end of the
+    // console when new content is added.
+    final scrollToEnd =
+        scrollInfo.top + scrollInfo.clientHeight >= scrollInfo.height;
+
+    final chunks = decodeAnsiColorEscapeCodes(text, _ansiUp);
+
+    final doc = _editor.getDoc();
+    for (StyledText chunk in chunks) {
+      final startPosition = _getDocumentEnd();
+      doc.replaceRange(chunk.text, startPosition);
+
+      final style = chunk.style;
+      if (style != null) {
+        doc.markText(startPosition, _getDocumentEnd(), css: style);
+      }
+    }
+
+    if (scrollToEnd) {
+      final documentEnd = _getDocumentEnd();
+      _editor.scrollIntoView(documentEnd.line, documentEnd.ch);
+    }
   }
 
-  String getContents() {
-    return _editor.getDoc().getValue();
+  /// This method returns a visualization of the contents of the console
+  /// suitable for integration tests.
+  ///
+  /// TextMarkers in the content are visualized with pseudo HTML+CSS to make it
+  /// easy to ensure that the styling of the content matches what was expected.
+  String getStyledContents() {
+    final doc = _editor.getDoc();
+    final marks = doc.getAllMarks();
+    final startTags = <Position, StringBuffer>{};
+    final endTags = <Position, StringBuffer>{};
+    for (var mark in marks) {
+      final positions = mark.find();
+      assert(positions.isNotEmpty);
+      final String css = mark.jsProxy['css'];
+      if (css == null || css.isEmpty) continue;
+      final start = positions.first;
+      final end = positions.last;
+      if (start.line != null && start.ch != null) {
+        final startTag = startTags.putIfAbsent(start, () => StringBuffer());
+        startTag.write("<span style='$css'>");
```

Comment:
[optional] consider skipping the otherwise unused local variable?

```dart
startTags.putIfAbsent(start, () => StringBuffer()).write(...);
```

Reply:
I think that makes it denser but a little harder to read. In this case I separate clearly what is the boilerplate getting the startTag from the map from the actually important part which is what is added as the start tag.
If it all fit in one line it might be fine but dartfmt makes this ugly. With a line length of 100 I would prefer the dartfmt version as it is at least all one line.
<img width="394" alt="Screen Shot 2019-08-24 at 2 05 21 PM" src="https://user-images.githubusercontent.com/1226812/63642716-5f5fcd80-c678-11e9-826a-3157ca7a0b35.png">